src/Utility/HttpResponsePaginator.php

```diff
@@ -0,0 +1,473 @@
+
+            // Is ?per_page present? If not, set a sane default.
+            if (! mb_strstr($params, 'per_page=')) {
+                $lastBuilder->withParam('per_page', 100);
+            }
+
+            // Ensure ?include_totals=true is present, required to iterate using this class.
+            $lastBuilder->withParam('include_totals', true);
+
+            // Increment our network request tracker for reference.
+            ++$this->requestCount;
+
+            // Issue next paged request.
+            try {
+                $lastBuilder->call();
+            } catch (\Auth0\SDK\Exception\NetworkException $exception) {
+                return false;
+            }
+
+            // Process the response.
+            return $this->processLastResponse($depth + 1);
+        }
+
+        return false;
+    }
+
+    /**
+     * Make a network request for the next page of results.
+     */
+    private function getNextResults(): bool
+    {
+        if ($this->lastBuilder() && $this->lastResponse()) {
+            // Retrieve the active HttpRequest instance to repeat the request.
+            $lastBuilder = $this->lastBuilder();
+
+            // Ensure basic pagination details are included in the request.
+            if ($this->usingCheckpointPagination) {
+                $lastBuilder->withParam('from', $this->nextCheckpoint);
+            } else {
+                // Get the next page.
+                $page = ceil($this->position / $this->requestLimit);
+
+                // Set the next page.
+                $lastBuilder->withParam('page', $page);
+            }
+
+            // Increment our network request tracker for reference.
+            ++$this->requestCount;
+
+            // Issue next paged request.
+            try {
+                $lastBuilder->call();
+            } catch (\Auth0\SDK\Exception\NetworkException $exception) {
+                return false;
+            }
+
+            // Process the response.
+            return $this->processLastResponse();
+        }
+
+        return false;
+    }
+
+    /**
+     * Process the previous HttpResponse results and cache them for iterator content.
+     */
+    private function processLastResponse(
+        int $depth = 0
+    ): bool {
+        $lastRequest = $this->lastRequest();
+        $lastResponse = $this->lastResponse();
+
+        if ($lastRequest && $lastResponse) {
+            // Was the HTTP request successful?
+            if (HttpResponse::wasSuccessful($lastResponse)) {
+                // Decode the response.
+                $results = HttpResponse::decodeContent($lastResponse);
+
+                $start = $results['start'] ?? null;
+
+                // No results, abort processing.
+                if (! is_array($results) || ! count($results)) {
+                    return false;
+                }
+
+                // There is no 'start' key, the request was probably made without the include_totals param. Try again using safe pagination defaults.
```

Comment:
I'm still not sure about this. If I make a request to some endpoint and choose to paginate it, would it not be reasonable to throw an error if the response doesn't contain the paging metadata? You also then wouldn't need to check if the url is a supported endpoint or not or risk retrying an endpoint that will never return pagination metadata.

```
if !results.start:
  throw "response not paginated, either the endpoint is not supported or you did not have the include_totals param"
```